Added error for publishing repo w/ non-existent org

Daniel-McCarthy opened the pull request:
As the docs say, open pull requests early. I figured would open this and see what others think.

This pull request is in response to Issue #3764. Although it does not cover all the suggestions in it, it is meant to solve the issue of the vague "Not Found" error when attempting to publish a repository with an organization that has been renamed or deleted. This change intends to make a very clear error that specifically tells the user that it is the organization that was not able to be found and shows the name of the organization to review.

I am certainly open to feedback and would be happy to make changes as needed. I am new both to this project and the conventions of this language, so I am sure there are better ways of doing things.

shiftkey commented on app/src/lib/api.ts:
```diff
@@ -374,6 +374,12 @@ export class API {
       return await parsedResponse<IAPIRepository>(response)
     } catch (e) {
       if (e instanceof APIError) {
+        const orgsList: ReadonlyArray<IAPIUser> | null = await this.fetchOrgs()
```
Do we need to re-fetch the organizations here? I think if we tried to `POST` to `orgs/${org.login}/repos` and got a 404 here it's reasonable to think this is related to the organization.

Daniel-McCarthy replied:
Wonderful, I was hoping there was a way to avoid fetching again. To clarify: if the org used is `orgs/${org.login}/repos` and received an APIError from this POST request then we have enough information to throw this error, and therefore no additional checks are needed?

shiftkey replied:
@Daniel-McCarthy yeah, I wouldn't worry about being cleverer here than that

shiftkey commented on app/src/lib/api.ts:
```diff
@@ -374,6 +374,12 @@ export class API {
+        const orgsList: ReadonlyArray<IAPIUser> | null = await this.fetchOrgs()
+        if (orgsList && org) {
+          if (orgsList.indexOf(org) === -1) {
+            throw new Error(`Organization "${org.login}" Not Found`)
```
We can iterate on this, but here's a suggestion for the error message itself:

> `Unable to create repository for organization '${org.login}' - check it exists and that you have permissions to create a repository there`

shiftkey commented on app/src/lib/api.ts:
```diff
@@ -374,6 +374,13 @@ export class API {
+          throw new Error(
+            `Unable to create repository for organization '${
+              org.login
+            }' - Verify it exists and that you have permission to create a repository there.`
```
:art: drop the `V` here from `Verify` or use a `.` instead of a `-` to indicate it's a new sentence.

Daniel-McCarthy replied:
Just to clarify, do you mean to change the V to not be capitalized? Would this be preferable to the period I have added in the most recent commit?

shiftkey commented on app/src/lib/api.ts:
```diff
@@ -374,6 +374,13 @@ export class API {
       return await parsedResponse<IAPIRepository>(response)
     } catch (e) {
       if (e instanceof APIError) {
+        if (org) {
```
:art: we're not fans of truthiness here - can you make this a bit more explicit by having a `if (org !== null) {` on this line?

Daniel-McCarthy commented:
Okay, I have updated again with the current set of suggestions. I made the null check more explicit to clarify the intent of it, and I have added a period to clarify the end of the first sentence. (I have kept the period outside of the quotations so that it can not be assumed to be part of the org name.)

Hopefully this fits better stylistically.

Daniel-McCarthy commented:
@shiftkey To clarify, this pull request is ready for review.

Should I clarify this for my other pull requests? I simply hope to get on the same page and do not want to come across as attempting to rush anything.

shiftkey commented:
One sec, confirming how the dialog looks when you trip this error (the new message has more words than the old message)...

shiftkey commented:
Flawless victory:

<img width="528" src="https://user-images.githubusercontent.com/359239/44670500-32db8c00-a9f9-11e8-985a-8fc91248d859.png">

billygriffin commented:
Thanks @Daniel-McCarthy, this is awesome! Excited to see it go live.
